src/smart-components/role/edit-resource-definitions-modal.js

```diff
@@ -0,0 +1,193 @@
+      name: 'dual-list-select',
+      leftTitle: 'Resources available for the permission',
+      rightTitle: 'Resources added to the permission',
+      filterOptionsTitle: 'Filter by resource...',
+      filterValueTitle: 'Filter by resource...',
+      options: [...(resourcesPath && resources ? options : [])],
+      validate: [{ type: 'validate-resources' }],
+    },
+  ],
+});
+
+const selector = ({ costReducer: { resourceTypes, isLoading, loadingResources, resources } }, resourcesPath) => ({
+  resourceTypes: resourceTypes.data,
+  resources: resources[resourcesPath] ? { resourcesPath: resources[resourcesPath] } : resources,
+  isLoading,
+  isLoadingResources: loadingResources > 0,
+});
+
+const validatorMapper = {
+  'validate-resources': () => (value) => (value && value.length > 0 ? undefined : 'At least one resource must be defined for this permission'),
+};
+
+const EditResourceDefinitionsModal = ({ cancelRoute }) => {
+  const routeMatch = useRouteMatch(paths['role-detail-permission-edit']);
+  const {
+    params: { permissionId, roleId },
+  } = useRouteMatch(routeMatch);
+  const { replace, push } = useHistory();
+
+  const dispatch = useDispatch();
+  const fetchResourceDefinitions = () => dispatch(getResourceDefinitions());
+
+  const [resourcesPath, setResourcesPath] = useState(undefined);
+  const [cancelWarningVisible, setCancelWarningVisible] = useState(false);
+  const [loadingStateVisible, setLoadingStateVisible] = useState(true);
+  const [changedResources, setChangedResources] = useState();
+
+  const { definedResources, role } = useSelector(
+    (state) => ({
+      role: state.roleReducer.selectedRole,
+      definedResources: state.roleReducer.selectedRole?.access
+        ? state.roleReducer.selectedRole.access
+            .find((a) => a.permission === permissionId)
+            .resourceDefinitions.reduce((acc, curr) => [...acc, curr.attributeFilter.value], []) // tady to jsem přehodil komentář, nefunguje!
+        : [],
+      isRecordLoading: state.roleReducer.isRecordLoading,
+    }),
+    shallowEqual
+  );
+
+  const { resourceTypes, isLoading, isLoadingResources, resources } = useSelector((props) => selector(props, resourcesPath), shallowEqual);
+
+  useEffect(() => {
+    fetchResourceDefinitions();
+  }, [permissionId]);
+
+  useEffect(() => {
+    if (!isLoading) {
+      let path = resourceTypes.find((r) => r.value === permissionId.split(':')?.[1])?.path;
+      if (path) {
+        setResourcesPath(path.split('/')[5]);
+        dispatch(getResource(path));
+      } else if (permissionId.split(':')?.[1] === '*') {
+        setResourcesPath('*');
+        resourceTypes.map((r) => dispatch(getResource(r.path)));
+      }
+    }
+  }, [resourceTypes]);
+
+  const onCancel = () => replace(cancelRoute);
+
+  const handleCancel = (data) => {
+    if (data['dual-list-select'] === definedResources) {
+      onCancel();
+    } else {
+      setChangedResources(data['dual-list-select']);
```

Comment:
Suggestion: multiple useState calls can be wrapped to one useReducer, it will prevent additional renders

Reply:
@rvsia Thanks for the advice! Just please take one more look if I am using it properly, thank you :)